Expand capabilities of the `debug` instruction

Overcastan opened the pull request:
This PR expands the capabilities of the `debug` instruction by printing information about memory and procedure locals, namely:
- `debug.mem` prints out the entire contents of RAM.
- `debug.mem.<n>` prints out contents of memory at address `n`.
- `debug.mem.<n>.<m>` prints out the contents of memory starting at address `n` and ending at address `m` (both inclusive). `m` must be greater than `n`.
- `debug.local` prints out all locals of the currently executing procedure.
- `debug.local.<n>` prints out contents of the local at index `n` for the currently executing procedure.
- `debug.local.<n>.<m>` prints out contents of the local starting at index `n` and ending at index `m` (both inclusive). `m` must be greater than `n`.

TODO:
- [ ] Update docs and changelog

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -39,14 +82,173 @@ fn print_vm_stack(clk: u32, stack: Vec<Felt>, n: usize) {
+
+    // print the main part of the memory (wihtout the last value)
+    for (addr, value) in mem.iter().take(mem.len() - 1) {
+        println!(
+            "├── {addr:>width$}: [{}, {}, {}, {}]",
+            value[0],
+            value[1],
+            value[2],
+            value[3],
+            width = padding
+        );
```
I would probably create a separate function to encapsulate this logic. Maybe something like:
```Rust
fn print_mem_address(clk: u32, ctx: u32, addr: u32, value: Option<Word>, is_last: bool)
```
Also, I wonder if we can just apply universal padding. Since memory addresses are 32 bits, we should fit into 8 hex characters (it is common to print memory address in hex).

We could probably have a similar function for locals (e.g., `print_loc_address`) with the main difference being that index of local should be printed out in decimal (rather than hex). Or we could combine the two functions and have an extra parameter to differentiate between the two.

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -39,14 +82,173 @@ fn print_vm_stack(clk: u32, stack: Vec<Felt>, n: usize) {
+    print_all: bool,
+) {
+    let mut local_mem_interval = Vec::new();
+    let local_memory_offset = 2u32.pow(30) + 1;
```
The offset should come from the `fmp` register. This way, it will be different for every procedure invocation. Otherwise, we'll be printing locals only for the first invoked procedure.

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -39,14 +82,173 @@ fn print_vm_stack(clk: u32, stack: Vec<Felt>, n: usize) {
+///
+/// `header` contains a title message corresponding to the DebugOption it is used in
+/// ([DebugOptions::MemInterval] or [DebugOptions::LocalInterval]).
+#[cfg(feature = "std")]
+fn print_mem_interval(mem: Vec<(u32, Option<Word>)>, header: String) {
```
To simplify things, I'd probably not have separate functions for handling individual address and intervals. We can probably treat a single address as a special case of intervals.

bobbinth commented on core/src/operations/decorators/debug.rs:
```diff
@@ -13,13 +13,40 @@ pub enum DebugOptions {
     StackAll,
     /// Prints out the top n items of the stack for the current context.
     StackTop(u16),
+    /// Prints out the entire contents of RAM.
+    MemAll,
+    /// Prints out contents of memory at the provided address.
+    MemAddr(u32),
+    /// Prints out the contents of memory stored in the provided interval. Interval boundaries are
+    /// both inclusive.
+    MemInterval(u32, u32),
+    /// Prints out contents of the local at the provided index for the currently executing
+    /// procedure.
+    LocalIndex(u32),
+    /// Prints out locals stored in the provided interval of the currently executing procedure.
+    /// Interval boundaries are both inclusive.
+    LocalInterval(u32, u32, bool),
+    /// Prints out the entire state of the VM (stack and RAM).
+    All(u16),
```
I wonder if we should reduce the number of ways we record this data. Similar to the previous comment, we can probably represent single values as special cases of intervals. Also, maybe we get rid of the `All` option for now. Basically, maybe the `DebugOptions` enum could look like this:
```Rust
pub enum DebugOptions {
    StackAll,
    StackTop(u16),
    MemAll,
    MemInterval(u32, u32),
    LocalInterval(u16, u16),
}
```
We'd still have all the previous options (except for the plain `debug`), we'd just translate `debug.mem.2` to `MemInterval(2, 2)` etc.

bobbinth reviewed (commented):
Thank you! Looks good! This is not a full review yet, but I wanted to leave a few comments and I think it will change things a bit.

Also, could you paste a couple of screen shots of how the results look like?

Overcastan commented:
There is a problem with formatting the final print of the memory address and value (as you mentioned it [here](https://github.com/0xPolygonMiden/miden-vm/pull/1103#discussion_r1359809168)). 

I didn't find the way to parametrize the output: in case we print in decimal we need to have something like `println!("{address:<10}: ...)` — so we want to add space for potentially long `u32` value and move it to the left side of the output. I also added dots to make it look a little bit more pretty (`println!("{address:.<10}: ...)`), but it is not important now. In case we want to print address in hex we use something like `println!("{address:#010x}: ...)` — `#x` for hex form and `010` for making hex to be length 10 which is spaced with 0's. The problem is that I don't know how to use some kind of parameter which would specify the output appearance (`.<10` in the first example and `#010x` in the second). 

Without that parametrization the code which outputs the values would be almost identical except the parts mentioned above, so there will be a lot of copy-paste. For now I left only the decimal form for both global and local memory. I am not sure how these addresses would be used in debugging, so I don't now how crucial or convenient will be using hexes instead of decimals. 

Now output looks like that: 

<img width="1132" alt="Screenshot 2023-10-16 at 10 03 54 PM" src="https://github.com/0xPolygonMiden/miden-vm/assets/43154072/ece25846-96d8-4697-bd53-b1868d5518ce">

bobbinth commented on processor/src/lib.rs:
```diff
@@ -551,6 +554,11 @@ pub trait ProcessState {
     /// been accessed previously.
     fn get_mem_value(&self, ctx: u32, addr: u32) -> Option<Word>;
+
+    /// Returns the entire memory state for the specified execution context at the current clock
+    /// cycle. The state is returned as a vector of (address, value) tuples, and includes addresses
+    /// which have been accessed at least once.
```
minor nit: I'd put a line break after the first sentence. This way, the docs will be formatted better. Specifically:
```
/// Returns the entire memory state for the specified execution context at the current clock
/// cycle.
///
/// The state is returned as a vector of (address, value) tuples, and includes addresses which
/// have been accessed at least once.
```

bobbinth commented on miden/examples/merkle_store/merkle_store.masm:
If we want to illustrate how debug instruction works, I'd probably create a separate example.

bobbinth commented on core/src/operations/decorators/debug.rs:
```diff
@@ -13,13 +13,35 @@ pub enum DebugOptions {
+    ///
+    /// Boolean parameter indicated whether whole local memory should be printed.
+    LocalInterval((u32, u32), u32, bool),
+    /// Prints out the entire state of the VM (stack and RAM).
+    All(u32),
```
Let's get rid of this option.

bobbinth commented on core/src/operations/decorators/debug.rs:
```diff
@@ -13,13 +13,35 @@ pub enum DebugOptions {
+    /// Prints out the contents of memory stored in the provided interval. Interval boundaries are
+    /// both inclusive.
+    MemInterval(u32, u32),
+    /// Prints out locals stored in the provided interval of the currently executing procedure.
+    /// Interval boundaries are both inclusive.
+    ///
+    /// Boolean parameter indicated whether whole local memory should be printed.
+    LocalInterval((u32, u32), u32, bool),
```
Let's simplify this and define it as `LocalInterval(u32, u32, u32)` - i.e., we can get rid of the boolean flag. Basically, if the interval covers all locals we can print it out as such.

also, we should explain what the 3 values are in the comments above.

Overcastan replied:
But won't it be confusing that the user has requested values in a certain interval, and he is given a smaller interval without any boundaries? Probably, in this case, we need to additionally specify the total amount of local memory, so it should be clearer.

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -1,52 +1,193 @@
-#[cfg(not(feature = "std"))]
-fn print_vm_stack(_clk: u32, _stack: Vec<Felt>, _n: usize) {
-    // in no_std environments, this is a NOOP
+fn print_word(addr: u32, value: &Word, last: bool) {
+    if last {
+        print!("└── ");
+    } else {
+        print!("├── ");
+    }
+    println!("{addr:.<10}: [{}, {}, {}, {}]", value[0], value[1], value[2], value[3],);
+    if last {
+        println!();
+    }
 }
```
I would replace this with:
```Rust
fn print_mem_address(addr: u32, value: Option<Word>, is_last: bool, is_local: bool) {

}
```
And handle all variations internally.

For memory address, you can use the following format `{add:#010x}`. For, local, you can use `{addr:>5}`.

Also, we should probably impose some width formatting for the values. How many characters does u64 require?

Overcastan replied:
u64 require 20 characters, but spacing each value with 20 white spaces will look strange. Here I can try to use the padding which depends on the maximum length of the value. In that case width of the output values will be equal, but potentially smaller than 20.

bobbinth replied:
Yep - I think that's a good idea!

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -1,52 +1,193 @@
+        }
     } else {
-        println!("├── {i:>2}: {}", stack[i]);
-        println!("└── ({} more items)", stack.len() - num_items);
+        println!("└── EMPTY\n");
     }
```
I don't think we need to print this out.

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -1,52 +1,193 @@
+        // print the last memory value
+        if let Some((addr, value)) = mem.last() {
+            print_word(*addr as u32, value, true);
+        } else {
+            println!("└── EMPTY\n");
+        }
```
I don't think we need to print this out.

bobbinth reviewed (changes requested):
Thank you! I left some more comments inline. Regarding formatting, I suggested how we can format addresses/indexes for mem vs. locals. Let me know if that doesn't work for some reason.

Overcastan commented:
I thought that it's strange to restrict user to request the locals in any interval he wants (as we do it in memory), so I get rid of the check if the end of the interval is smaller than number of locals. Instead I added a check that the end of the interval is smaller than 2^16 (since it's forbidden to have more than 2^16 locals for each procedure). Using it I also implemented the "print whole local memory" variant without boolean flag, using 2^16 as an end of the interval in that case. 

What do you think about this decision? Should I return everything as it was?

Now output looks like this:

<img width="1036" alt="Screenshot 2023-10-17 at 7 18 53 PM" src="https://github.com/0xPolygonMiden/miden-vm/assets/43154072/d0f623b8-3e51-4c5e-9f46-14cb9f61c7de">

bobbinth commented on core/src/operations/decorators/debug.rs:
```diff
@@ -13,13 +13,31 @@ pub enum DebugOptions {
+    ///
+    /// First parameter specifies the starting address, second -- the ending address, and the third
+    /// specifies the overall number of locals.
+    LocalInterval(u32, u32, u32),
```
Number of locals should not be greater than `u16` - so, we should probably use `u16`'s here instead of `u32`s.

bobbinth commented on assembly/src/ast/parsers/debug.rs:
```diff
@@ -29,6 +29,38 @@ pub fn parse_debug(op: &Token) -> Result<Node, ParsingError> {
+                if m < n {
+                    return Err(ParsingError::invalid_param_with_reason(op, 3, "the index of the end of the interval must be greater than the index of its beginning"));
+                }
+                DebugOptions::LocalInterval(n, m, 0)
```
Rather than setting number of local constants to 0 here and then updating later, we can probably pass the number of local constants into this function (similar to how we do it for `parse_loc_*` functions.

Overcastan replied:
I'm not sure I understand what do you mean. As far as I know `local_constants` contains only constants declared with `CONST` word, so number of locals is not stored there. It is stored in the `local_procs: LocalProcMap` (in `ProcedureAst` [here](https://github.com/0xPolygonMiden/miden-vm/blob/775209f1dd0b3903a870f22cb2b804347697d439/assembly/src/ast/mod.rs#L740)), but this information becomes available only after the procedure is completely parsed, so we can't get `num_locals` being within parsed procedure. That's why I had to create this ugly solution with updating already created `DebugOptions::LocalInterval` at the compilation stage, when we already have this data in the AssemblyContext. 

I hope that I just didn't understand something and there is a way to get number of locals at the parsing stage to avoid modifying `DebugOptions` later.

bobbinth replied:
Ah, sorry! You are correct - we currently don't have this info. But looking at the code again, it actually doesn't seem like it should be too difficult to track it.

What we can do is add `num_proc_locals: Option<u16>` to the [ParserContext](https://github.com/0xPolygonMiden/miden-vm/blob/main/assembly/src/ast/parsers/context.rs#L13). Then, in [parse_procedure()](https://github.com/0xPolygonMiden/miden-vm/blob/main/assembly/src/ast/parsers/context.rs#L292) you could set it to the number of locals right before we parse the procedure body, and then reset it back to `None` after procedure body is parsed.

We can probably use this info in other places as well - but that's for a different PR.

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -1,52 +1,242 @@
+) {
+    if let Some(value) = value {
+        if is_last {
+            if is_local {
+                println!(
+                    "└── {addr:>5}: [{:>width$}, {:>width$}, {:>width$}, {:>width$}]\n",
+                    value[0].as_int(),
+                    value[1].as_int(),
+                    value[2].as_int(),
+                    value[3].as_int(),
+                    width = padding
+                )
+            } else {
+                println!(
+                    "└── {addr:#010x}: [{:>width$}, {:>width$}, {:>width$}, {:>width$}]\n",
+                    value[0].as_int(),
+                    value[1].as_int(),
+                    value[2].as_int(),
+                    value[3].as_int(),
+                    width = padding
+                )
+            }
```
I'd probably do this like so:
```Rust
if is_local {
    print!("└── {addr:>5}: ");
} else {
    print!("└── {addr:#010x}: ");
}
print_word(value, padding);
println!();
```
Where `print_word()` is a new function which uses `println!` to print out the word.

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -1,52 +1,242 @@
+                    width = padding
+                )
+            }
+        } else if is_local {
+            println!(
+                "├── {addr:>5}: [{:>width$}, {:>width$}, {:>width$}, {:>width$}]",
+                value[0].as_int(),
+                value[1].as_int(),
+                value[2].as_int(),
+                value[3].as_int(),
+                width = padding
+            )
+        } else {
+            println!(
+                "├── {addr:#010x}: [{:>width$}, {:>width$}, {:>width$}, {:>width$}]",
+                value[0].as_int(),
+                value[1].as_int(),
+                value[2].as_int(),
+                value[3].as_int(),
+                width = padding
+            )
+        }
```
And this can also probably be simplified by using `print_word()` function.

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -1,52 +1,242 @@
+/// Returns the maximum length among the word elements.
+fn word_elem_max_len(word: Option<Word>) -> usize {
+    if let Some(word) = word {
+        word.iter().fold(0, |max, value| value.as_int().to_string().len().max(max))
```
Rather than converting to string and counting characters, wouldn't [ilog10()](https://doc.rust-lang.org/std/primitive.u64.html#method.ilog10) give us the same result?

bobbinth reviewed (changes requested):
Looks good! Thank you! I left a few more comments inline.

bobbinth commented:
> What do you think about this decision?

I think it looks good! Let's keep it as you have it.

bobbinth commented on assembly/src/assembler/instruction/mod.rs:
```diff
@@ -408,3 +408,18 @@ where
     })
 }
+
+// / Returns a [Decorator::Debug] instance created with information about number of procedure
+// / locals.
+// fn process_debug_options(
+//     ctx: &AssemblyContext,
+//     options: &DebugOptions,
+// ) -> Result<Decorator, AssemblyError> {
+//     let num_locals = ctx.num_proc_locals();
+//     match options {
+//         DebugOptions::LocalInterval(start, end, _) => {
+//             Ok(Decorator::Debug(DebugOptions::LocalInterval(*start, *end, num_locals)))
+//         }
+//         _ => Ok(Decorator::Debug(*options)),
+//     }
+// }
```
This can probably be removed.

bobbinth commented on assembly/src/ast/parsers/context.rs:
```diff
@@ -15,6 +15,7 @@ pub struct ParserContext<'a> {
     pub local_procs: LocalProcMap,
     pub reexported_procs: ReExportedProcMap,
     pub local_constants: LocalConstMap,
+    pub num_procs_local: u16,
```
nit: I would name this `num_proc_locals`.

bobbinth commented on assembly/src/ast/parsers/debug.rs:
```diff
@@ -14,7 +14,7 @@ use vm_core::DebugOptions;
 /// Returns an error if the instruction token contains a wrong number of parameters, or if
 /// the provided parameters are not valid.
-pub fn parse_debug(op: &Token) -> Result<Node, ParsingError> {
+pub fn parse_debug(op: &Token, num_procs_local: u16) -> Result<Node, ParsingError> {
```
nit: I'd rename `num_procs_local` into `num_proc_locals`.

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -1,52 +1,234 @@
+        }
+
+        if interval.0 == 0 && interval.1 == u16::MAX as u32 {
+            println!("Local state before step {} for the context {}:", self.clk, self.ctx)
```
I would change the printed out text to: "State of procedure locals before step {}:".

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -1,52 +1,234 @@
+            println!("Local state before step {} for the context {}:", self.clk, self.ctx)
+        } else if interval.0 == interval.1 {
+            println!(
+                "Local state before step {} for the context {} at index {}:",
```
I would change the printed out text to: "State of procedure local at index {} before step {}:".

bobbinth commented on processor/src/host/debug.rs:
```diff
@@ -1,52 +1,234 @@
+            )
+        } else {
+            println!(
+                "Local state before step {} for the context {} in the interval [{}, {}]:",
```
I would change the printed out text to: "State of procedure locals [{}, {}] before step {}:".

bobbinth reviewed (approved):
Looks good! Thank you! I left a few nit comments inline. After these are addressed, we can merge.
